lib/matplotlib/colors.py

```diff
@@ -1225,6 +1240,594 @@ def reversed(self, name=None):
+            interval ``[0, 1]`` otherwise they will be `numpy.uint8`\s in the
+            interval ``[0, 255]``.
+
+        Returns
+        -------
+        Tuple of RGBA values if X[0] is scalar, otherwise an array of
+        RGBA values with a shape of ``X.shape + (4, )``.
+        """
+
+        if len(X) != len(self):
+            raise ValueError(
+                f'For the selected colormap the data must have a first dimension '
+                f'{len(self)}, not {len(X)}')
+        rgba, mask_bad = self[0](X[0], bytes=False, return_mask_bad=True)
+        rgba = np.asarray(rgba)
+        for c, xx in zip(self[1:], X[1:]):
+            sub_rgba, sub_mask_bad = c(xx, bytes=False, return_mask_bad=True)
+            sub_rgba = np.asarray(sub_rgba)
+            rgba[..., :3] += sub_rgba[..., :3]  # add colors
+            rgba[..., 3] *= sub_rgba[..., 3]  # multiply alpha
+            mask_bad |= sub_mask_bad
+
+        if self.combination_mode == 'Sub':
+            rgba[..., :3] -= len(self) - 1
+
+        rgba[mask_bad] = self.get_bad()
+
+        rgba = np.clip(rgba, 0, 1)
+
+        if alpha is not None:
+            alpha = np.clip(alpha, 0, 1)
+            if alpha.shape not in [(), np.array(X[0]).shape]:
+                raise ValueError(
+                    f"alpha is array-like but its shape {alpha.shape} does "
+                    f"not match that of X[0] {np.array(X[0]).shape}")
+            rgba[..., -1] *= alpha
+
+        if bytes:
+            rgba = (rgba * 255).astype('uint8')
+
+        if not np.iterable(X[0]):
+            rgba = tuple(rgba)
+
+        return rgba
+
+    def copy(self):
+        """Return a copy of the multivarcolormap."""
+        return self.__copy__()
+
+    def __copy__(self):
+        cls = self.__class__
+        cmapobject = cls.__new__(cls)
+        cmapobject.__dict__.update(self.__dict__)
+        cmapobject.colormaps = [cm.copy() for cm in self.colormaps]
+        cmapobject._rgba_bad = np.copy(self._rgba_bad)
+        return cmapobject
+
+    def __eq__(self, other):
+        if not isinstance(other, MultivarColormap):
+            return False
+        if not len(self) == len(other):
+            return False
+        for c0, c1 in zip(self, other):
+            if not c0 == c1:
+                return False
+        if not all(self._rgba_bad == other._rgba_bad):
+            return False
+        if not self.combination_mode == other.combination_mode:
+            return False
+        return True
+
+    def __getitem__(self, item):
+        return self.colormaps[item]
+
+    def __iter__(self):
+        for c in self.colormaps:
+            yield c
+
+    def __len__(self):
+        return len(self.colormaps)
+
+    def __str__(self):
+        return self.name
+
+    def get_bad(self):
+        """Get the color for masked values."""
+        return np.array(self._rgba_bad)
+
+    def set_bad(self, color='k', alpha=None):
```

Comment:
I suggest to make the new colormaps immuatble. This should have been the case with current colormaps as well. Instead of modifying the existing colormap, return a new colormap with modified properties; c.f. `Colormap.with_extremes`, https://github.com/matplotlib/matplotlib/pull/14645#issuecomment-506958221

Reply:
This is reasonable suggestion, but there is a practical complication. 
`MultivarColormap` contains a list of `Colormap` objects, which themselves are not immutable. 
(I implemented `MultivarColormap.__getitem__()` as a way to get the individual constituent colormaps, because I believe this is useful functionality.) 
If we make `MultivarColormap` immutable while the constituent colormaps are not I'm afraid we will run into some trouble in this regard.

Reply:
Agreed, this is not optimal. But it's not as bad as one may think. Since colormaps are not immuatable `colormaps[name]` or `get_cmap(name)` return a new copy to prevent users from modifying the builtin colormaps. This also means, if you provide these colormaps as input to `MultivarColormap` and don't keep references, you're safe. Additionally, I find it important from an education perspective to not give the users the tools to modify existing colormaps, even if we cannot ensure full immutability. And finally, we may be able to move towards immutable colormaps one day and thus should not put additional obstacles in our way.

Reply:
Just so we are clear, this is the kind of use case I am envisioning:
```
cmaps = matplotlib.multivar_colormaps['2VarAddA']
norm_growth = matplotlib.colors.Normalize(vmin=2, vmax=10)
norm_capita = matplotlib.colors.Normalize(vmin=50, vmax=110)

p0 = matplotlib.collections.PatchCollection(patches, cmap=cmaps[0], norm=norm_capita)
p1 = matplotlib.collections.PatchCollection(patches, cmap=cmaps[1], norm=norm_growth)
p2 = matplotlib.collections.PatchCollection(patches, cmap=cmaps,  norm=(norm_capita, norm_growth))
...
```
![image](https://github.com/matplotlib/matplotlib/assets/2361452/f315854d-df8e-4d74-b891-3773cc11a4bb)

Where two datasets are shown independently, and also together, so that correlations can (ideally) be more easily identified. In order to use the same colormap both to visualize data independently, and together the constituent `Colormap` objects of the `MultivarColormap` object needs to accessed, i.e. via:
```
    def __getitem__(self, item):
        return self.colormaps[item]
```
Are you thinking we should change this to:
```
    def __getitem__(self, item):
        return self._colormaps[item].copy()
```
Where the member variable has been made private and we only return a copy.
With these changes we can then make the `MultivarColormap` objects immutable.

(PS: I would like to use the above figure as part of an example in a later PR)

Reply:
Not exactly what I meant. I'm merely suggesting not to add any in-place modifying functions.

The copy comments were more to explain what happens with colormaps currently, and that you be don't have to be too concerned that somebody changes the input s you use under you. I haven't thought about extracting colormaps from the multivar colormaps. In particular here, I'd not make a copy because the colormaps are logically linked.

On a general note, I find this example difficult to follow; e.g. Washington is light grey. That this is a combination of saturated blue and saturated brown is hard to realize. This multivar approach has the fundamental problem that it can produce colors that are far from participating colormaps. The degree of the problem depends on the data and on the chosen colormaps. I feel for this visualization a bivar mapping like ![image](https://github.com/matplotlib/matplotlib/assets/2836374/ef364bc6-13b1-4922-be99-19bcbfe5662d) is more suited. Here one can easily see: black=both low, white=both high, orange/blue=one high and one low.

Reply:
This discussion got a bit sidetracked. 
Regarding the original question, I removed all `set_` functions and introuduced `with_extremes()` for both BivarColormap and MultivarColormap.